pkg/kubefed2/federate/util.go

```diff
@@ -49,3 +59,126 @@ func SetBasicMetaFields(resource *unstructured.Unstructured, apiResource metav1.
 		resource.SetNamespace(namespace)
 	}
 }
+
+func namespacedAPIResourceMap(config *rest.Config, skipAPIResourceNames string) (map[string]metav1.APIResource, error) {
+	apiResourceLists, err := enable.GetServerPreferredResources(config)
+	if err != nil {
+		return nil, err
+	}
+
+	apiResources := make(map[string]metav1.APIResource)
+	for _, apiResourceList := range apiResourceLists {
+		if len(apiResourceList.APIResources) == 0 {
+			continue
+		}
+
+		gv, err := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+		if err != nil {
+			return nil, errors.Wrap(err, "Error parsing GroupVersion")
+		}
+
+		group := gv.Group
+		for _, apiResource := range apiResourceList.APIResources {
+			if !apiResource.Namespaced || isFederatedAPIResource(apiResource.Kind, group) ||
+				apiResourceMatchesSkipNames(apiResource, skipAPIResourceNames, group) {
+				continue
+			}
+
+			if previousAPIResource, ok := apiResources[apiResource.Kind]; ok {
+				if compareGroupVersions(gv, schema.GroupVersion{Group: previousAPIResource.Group, Version: previousAPIResource.Version}) <= 0 {
+					// The newer GV is not latest keep the previous.
+					continue
+				}
+			}
+
+			// The individual apiResources do not have the group and version set
+			apiResource.Group = group
+			apiResource.Version = gv.Version
+
+			apiResources[apiResource.Kind] = apiResource
+		}
+	}
+
+	return apiResources, nil
+}
+
+func compareGroupVersions(newGV, oldGV schema.GroupVersion) int {
+	// The strategy involved to choose a Group higher in order here is to
+	// consider 'extensions' as the outdated group [This seems to be true for
+	// all k8s APIResources, so far]. For example deployments exists in
+	// 'extensions' and 'apps'; 'deployments.apps' will be chosen.
+	// This doesn't apply to events but events are listed in controllerCreatedAPIResourceNames
+	// and skipped always.
+	// TODO(irfanurrehman) Are there any other known groups which don't match this?
+	if oldGV.Group != newGV.Group {
```

Comment:
1. Before implementing this, I did go through the list of k8s APIs and found the below to be available across more then 1 group:
```
events      events.k8s.io
events       "" (core)

daemonsets   apps
daemonsets   extensions 
deployments   apps 
deployments   extensions 
replicasets      apps 
replicasets      extensions

networkpolicies   extensions 
networkpolicies   networking.k8s.io 

podsecuritypolicies  extensions   
podsecuritypolicies  policy 
 ```
Except `events` all others were earlier in `extensions` but evolved into their own groups; so the strategy chosen applies (Please correct me if I am wrong). 
However it seems logical to add code to error out in case any other API (say a CRD in future) which does not map to this strategy is discovered withing multiple groups. Updated to this effect.

> Note that in the case of extensions and apps, only apps has `v1`, so a straight version comparison might suffice.

2. Does not seem like the right strategy.. :).  Some more context -> We use `ServerPreferredResources` API on `discovery client` to get all API resources, which returns one version per group in almost all scenario. The code does not reach `CompareKubeAwareVersionStrings` thus. However I have still used the version comparison because `ServerPreferredResources` does not guarantee that it will always have a preferred version available for a given API resource.